freqtrade/optimize/hyperopt_loss_calmar.py

```diff
@@ -0,0 +1,50 @@
+"""
+CalmarHyperOptLoss
+
+This module defines the alternative HyperOptLoss class which can be used for
+Hyperoptimization.
+"""
+from datetime import datetime
+
+import numpy as np
+from freqtrade.data.btanalysis import calculate_max_drawdown
+from freqtrade.optimize.hyperopt import IHyperOptLoss
+from pandas import DataFrame
+
+
+class CalmarHyperOptLoss(IHyperOptLoss):
+    """
+    Defines the loss function for hyperopt.
+
+    This implementation uses the Calmar Ratio calculation.
+    """
+
+    @staticmethod
+    def hyperopt_loss_function(results: DataFrame, trade_count: int,
+                               min_date: datetime, max_date: datetime,
+                               *args, **kwargs) -> float:
+        """
+        Objective function, returns smaller number for more optimal results.
+
+        Uses Calmar Ratio calculation.
+        """
+        total_profit = results["profit_ratio"]
+        days_period = (max_date - min_date).days
+
+        # adding slippage of 0.1% per trade
+        total_profit = total_profit - 0.0005
+        expected_returns_mean = total_profit.sum() / days_period
+
+        # calculate max drawdown
+        try:
+            _,_,_,high_val,low_val = calculate_max_drawdown(results)
```

Comment:
Please be careful with the `max_drawdown()` calculation returned by this method.

it's "per stake" - not the overall total drawdown (maybe read through #5167 - which should explain the issue ...)

Reply:
Also - you should run `flake8` locally to have it pass right away ... ;)

Reply:
thank you for the insight. I was testing the hyperopt right now, and "profit_abs" is the way to go. I made the change to line 40 to  "_,_,_, high_val, low_val = calculate_max_drawdown(results, value_col='profit_abs')"  which gave me the max drawdown ratio.

Reply:
Calmar Ratio = Total Profit % / Max Drawdown %

I am curious on how I should approach the total profit variable on line 32. For example, let's say the total profit is 203%. When I try using the "total_profit = results['profit_ratio'].sum()" it gives me the decimal 20.30 which equals to 2030%, and if I use "total_profit = results['profit_abs'].sum()" it gives me the dollar amount 20300USDT. So I am having trouble getting the correct total profit. I think maybe importing the starting balance could help me fix my issue because I could then use this formula "total_profit = results['profit_abs'].sum() / starting_balance" which will give me the correct decimal to equal 203%. I am worried it will slow down the hyperopt file though. How would you recommend that I get the total profit percentage?

Reply:
Assuming you're using compounding (stake_amount=unlimited) - then you can't use `sum(profit_ratio)` - as it's not a weighted sum.

The last parameter of the hyperopt-loss function is `backtest_stats` - (assuming you're using the one from [here](https://www.freqtrade.io/en/stable/advanced-hyperopt/#creating-and-using-a-custom-loss-function)).

While it's contents are not very well documented, it contains everything backtesting is showing - so you can simply use the proper value from there (most likely `backtest_stats['profit_total']` - which contains the ratio - multiply with 100 to get %).

Reply:
That did the job! I'm pretty confident I resolved this issue :D